### PR TITLE
Bump @babel/preset-env to 7.8.3 (Optional Chaining!)

### LINKED ISSUE #18939: Babel Preset: Update @babel/preset-env to enabled optional chaining
Optional chaining reached stage 4!

image

https://twitter.com/NicoloRibaudo/status/1202323456899067906

This means that... it should be soon enabled by default in `@babel/preset-env` and in `@babel/parser`.

There are two issues to track progress:
- https://github.com/babel/babel/issues/10809
- https://github.com/babel/babel/issues/10810

Once those two packages are updated and published to npm we should take an action and update `@wordpress/babel-preset-default` and all other packages using Babel to use the latest version of all Babel packages.

## Resources

- https://tc39.es/proposal-optional-chaining/
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining
- https://2ality.com/2019/07/optional-chaining.html

### PR DESCRIPTION
**Pull request by sainthkh**
- Closes #18939
- #19065 should be merged, too, for `shippedProposals`. 

## Description

Bump @babel/preset-env to 7.8.3. Now we can use Optional Chaining!

## How has this been tested?

Added test code below to unit test:

```js
test( 'support for optional chaining', () => {
	const obj = {
		foo: {
			bar: 42,
		},
	};

	expect( obj?.foo?.bar ).toEqual( 42 );
	expect( obj?.foo?.baz ).toEqual( undefined );
} );
```

## Screenshots

N/A

## Types of changes

New Feature for Developers

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/babel-preset-default/package.json [hunk 1]
```diff
@@ -27,13 +27,13 @@
 	],
 	"main": "index.js",
 	"dependencies": {
-		"@babel/core": "^7.4.5",
+		"@babel/core": "^7.8.3",
 		"@babel/plugin-proposal-async-generator-functions": "^7.2.0",
```
Comment (gziolo) [inline packages/babel-preset-default/package.json]: It would be great to update other Babel plugins to their latest version for consistency and the best experience.

**Review by gziolo — COMMENTED** [review]
Awesome work, thanks for opening this patch. I'm looking forward to using new language features 💯 

Can we add some more changes to ensure that all WordPress packages and Gutenberg use the latest version of Babel packages?

Some places where Babel packages are in use:

https://github.com/WordPress/gutenberg/blob/c079120ce39044e53ea41a6a1a246aeec8195433/package.json#L75-L78

https://github.com/WordPress/gutenberg/blob/c079120ce39044e53ea41a6a1a246aeec8195433/package.json#L108-L113

Besides, the majority of plugins ship with `@babel/runtime` as a dependency, e.g.:
https://github.com/WordPress/gutenberg/blob/5a8eda1d91bdde01d6bb5b2684b94339094b7235/packages/a11y/package.json#L25

The last part is updating the CHANGELOG for `@wordpress/babel-preset-default`  – I can help with that part if you have any concerns. Related file:
https://github.com/WordPress/gutenberg/blob/f841510811b619af9526b1e26af41bdf26d56314/packages/babel-preset-default/CHANGELOG.md
related docs:
https://github.com/WordPress/gutenberg/tree/f841510811b619af9526b1e26af41bdf26d56314/packages#maintaining-changelogs

File: packages/babel-preset-default/package.json (hunk 1 above)
Comment (sainthkh) [reply packages/babel-preset-default/package.json]: There were `7.4.4` and `7.other`. So, I thought changing them would be sufficient. I'll fix them soon.

Comment (gziolo) [reply packages/babel-preset-default/package.json]: if you encounter any issues we can split this into 2 PRs.

**Review by gziolo — APPROVED** [review]
Excellent, this looks great. I will wait for Travis to make sure it all works properly with the latest version of Babel packages. Thank you for applying changes so quickly. Should I wait for CHANGELOG.md changes or cover it myself (we mostly care about the preset-env as it adds new exciting features)?

**Review by gziolo — CHANGES_REQUESTED** [review]
I was too fast with approval.

Travis reports errors in `npm run lint-js` execution:

https://travis-ci.com/WordPress/gutenberg/jobs/279206011#L396

I guess we need to bump some ESLint dependencies responsible for Babel integration.

**Comment by sainthkh** [comment]
@gziolo 

1. Regenerated `package-lock.json` from scratch because that lint null bug was caused by conflicting dependencies. 
2. Had to remove ^ from `elements` package because it makes it use `react@16.12.0`. And it caused conflicts. That caused all the test failures.
3. Fixed snapshot because re-resizable is updated to 6.1.1
4. Manually downgraded xmldom because it caused license error. 

Now, all test passed. Please let me know if there is any problem.

**Comment by gziolo** [comment]
I extended the changelog entry in: 2a27bd9.

I tried applying the `package-lock.json` in 3a05bf0 but it looks like doing it through web interface doesn't quite work as expected. This is what I did:
- I aded back ranges to the `react` and `reac-dom` dependencies in `@wordpress/elements`
- reverted all changes in the lock file
- removed manually all `@babel/*` dependencies listed at the beginning of the lock file
- run `npm install`

Basically this ensures that only Babel dependencies are updated and leaves everything else in the same shape.

**Comment by gziolo** [comment]
I reverted changes in the `@wordpress/element` and the snapshot file you mentioned. All this should be enough to make all Travis checks pass but the one which checks the lock file. Let's see :)


You are doing an amazing job, the lock file is the most annoying part of the monorepo setup!

**Comment by gziolo** [comment]
I think I sorted it out, please look at https://github.com/WordPress/gutenberg/pull/19859. Can you remove 3 last commits and cherry-pick 55d7785fe653a57a0bca343123ff51f729b302a3? This way we can make sure that all the work will be committed under your name :)

**Comment by gziolo** [comment]
@sainthkh you rock, thanks for so quick respond. I'm going to merge it as soon as Travis is green. In my branch, it is 🎉

**Review by gziolo — APPROVED** [review]
Travis is more than happy with the changes proposed 😄

**Comment by sainthkh** [comment]
@gziolo Thanks for all the effort!
